packages/user-operation-controller/src/UserOperationController.test.ts

```diff
@@ -0,0 +1,450 @@
+        { ...ADD_USER_OPERATION_OPTIONS_MOCK, smartContractAccount },
+      );
+
+      await hash();
+
+      expect(bundlerMock.estimateUserOperationGas).toHaveBeenCalledTimes(1);
+      expect(bundlerMock.estimateUserOperationGas).toHaveBeenCalledWith(
+        {
+          callData: PREPARE_USER_OPERATION_RESPONSE_MOCK.callData,
+          callGasLimit: '0x1',
+          initCode: PREPARE_USER_OPERATION_RESPONSE_MOCK.initCode,
+          maxFeePerGas: ADD_USER_OPERATION_REQUEST_MOCK.maxFeePerGas,
+          maxPriorityFeePerGas:
+            ADD_USER_OPERATION_REQUEST_MOCK.maxPriorityFeePerGas,
+          nonce: PREPARE_USER_OPERATION_RESPONSE_MOCK.nonce,
+          paymasterAndData:
+            PREPARE_USER_OPERATION_RESPONSE_MOCK.dummyPaymasterAndData,
+          preVerificationGas: '0x1',
+          sender: PREPARE_USER_OPERATION_RESPONSE_MOCK.sender,
+          signature: PREPARE_USER_OPERATION_RESPONSE_MOCK.dummySignature,
+          verificationGasLimit: '0x1',
+        },
+        ENTRYPOINT,
+      );
+
+      expect(bundlerMock.sendUserOperation).toHaveBeenCalledTimes(1);
+      expect(bundlerMock.sendUserOperation).toHaveBeenCalledWith(
+        expect.objectContaining({
+          callGasLimit: '0xb9',
+          preVerificationGas: '0x2ac',
+          verificationGasLimit: '0x4a0',
+        }),
+        ENTRYPOINT,
+      );
+    });
+
+    it('marks user operation as failed if error', async () => {
+      const controller = new UserOperationController({
+        messenger,
+      });
+
+      const error = new Error(ERROR_MESSAGE_MOCK);
+      (error as any).code = ERROR_CODE_MOCK;
+
+      bundlerMock.sendUserOperation.mockRejectedValue(error);
+
+      const { id, hash } = await controller.addUserOperation(
+        ADD_USER_OPERATION_REQUEST_MOCK,
+        { ...ADD_USER_OPERATION_OPTIONS_MOCK, smartContractAccount },
+      );
+
+      await expect(hash()).rejects.toThrow(ERROR_MESSAGE_MOCK);
+
+      expect(Object.keys(controller.state.userOperations)).toHaveLength(1);
+      expect(controller.state.userOperations[id]).toStrictEqual(
+        expect.objectContaining({
+          error: {
+            code: ERROR_CODE_MOCK,
+            message: ERROR_MESSAGE_MOCK,
+            name: error.name,
+            rpc: undefined,
+            stack: error.stack,
+          },
+          id,
+          status: UserOperationStatus.Failed,
+        }),
+      );
+    });
+
+    // eslint-disable-next-line jest/expect-expect
+    it('does not throw if hash function not invoked', async () => {
+      const controller = new UserOperationController({
+        messenger,
+      });
+
+      bundlerMock.sendUserOperation.mockRejectedValue(
+        new Error(ERROR_MESSAGE_MOCK),
+      );
+
+      await controller.addUserOperation(ADD_USER_OPERATION_REQUEST_MOCK, {
+        ...ADD_USER_OPERATION_OPTIONS_MOCK,
+        smartContractAccount,
+      });
+
+      await new Promise((resolve) => setImmediate(resolve));
```

Comment:
We could, but the syntax is a little awkward given it's async code, so I'd argue it's cleanest to just assume any unhandled exception would fail the test.

Have added a code comment to clarify that.